great_expectations/expectations/metrics/column_aggregate_metrics/column_distinct_values.py

```diff
@@ -42,49 +54,67 @@ def _spark(
         runtime_configuration: Dict,
-    ):
-        observed_value_counts = metrics["column.value_counts"]
-        return set(observed_value_counts.index)
-
-    @classmethod
-    def _get_evaluation_dependencies(
-        cls,
-        metric: MetricConfiguration,
-        configuration: Optional[ExpectationConfiguration] = None,
-        execution_engine: Optional[ExecutionEngine] = None,
-        runtime_configuration: Optional[Dict] = None,
-    ):
-        """Returns a dictionary of given metric names and their corresponding configuration,
-        specifying the metric types and their respective domains"""
-        dependencies: dict = super()._get_evaluation_dependencies(
-            metric=metric,
-            configuration=configuration,
-            execution_engine=execution_engine,
-            runtime_configuration=runtime_configuration,
+    ) -> Set[Any]:
+        df, _, accessor_domain_kwargs = execution_engine.get_compute_domain(
+            metric_domain_kwargs, MetricDomainTypes.COLUMN
+        )
+        column = accessor_domain_kwargs["column"]
+
+        distinct_values = (
+            df.select(column)
+            .distinct()
+            .where(F.col(column).isNotNull())
+            .rdd.flatMap(lambda x: x)
+            .collect()
         )
 
-        if isinstance(
-            execution_engine, (SqlAlchemyExecutionEngine, SparkDFExecutionEngine)
-        ):
-            dependencies["column.value_counts"] = MetricConfiguration(
-                metric_name="column.value_counts",
-                metric_domain_kwargs=metric.metric_domain_kwargs,
-                metric_value_kwargs={
-                    "sort": "value",
-                    "collate": None,
-                },
-            )
-
-        return dependencies
+        return set(distinct_values)
 
 
 class ColumnDistinctValuesCount(ColumnAggregateMetricProvider):
     metric_name = "column.distinct_values.count"
 
     @column_aggregate_value(engine=PandasExecutionEngine)
-    def _pandas(cls, column, **kwargs):
+    def _pandas(cls, column, _metrics, **kwargs):
+        if "column.distinct_values" in _metrics:
+            return len(_metrics["column.distinct_values"])
         return column.nunique()
 
+    @column_aggregate_partial(engine=SqlAlchemyExecutionEngine)
+    def _sqlalchemy(cls, column, _metrics, **kwargs):
+        if "column.distinct_values" in _metrics:
+            return sa.literal(len(_metrics["column.distinct_values"]))
+        return sa.func.count(column.distinct())
+
+    @column_aggregate_partial(engine=SparkDFExecutionEngine)
+    def _spark(
+        cls,
+        column,
+        _metrics,
+        **kwargs,
+    ):
+        if "column.distinct_values" in _metrics:
+            return F.lit(len(_metrics["column.distinct_values"]))
+        return F.countDistinct(column)
+
+
+class ColumnDistinctValuesCountUnderThreshold(ColumnAggregateMetricProvider):
+    metric_name = "column.distinct_values.count.under_threshold"
```

Comment:
@NathanFarmer Would you be opposed to a slight renaming -- to "column.distinct_values.count_under_threshold" (replacing the last period with the underscore)?  Thanks!

Reply:
The only issue I have with this is that it doesn't align to `column_values.z_score.under_threshold`. This depends on `column.distinct_values.count` so shouldn't we communicate that information in the name?

Reply:
@NathanFarmer I see -- got it (agreed).  Thanks.